src/compute/src/sink/persist_sink.rs

```diff
@@ -1209,3 +1218,114 @@ where
+    let mut input = builder.new_input_connection(&coll.inner, Pipeline, vec![Antichain::new()]);
+    builder.build(move |capabilities| {
+        // This capability directly controls this operator's output frontier (because we have disconnected the input
+        // above). We wrap it in an Option so we can drop it to advance to the empty output frontier when the last
+        // refresh is done. (We must be careful that we only ever emit output updates at times that are at or beyond
+        // this capability.)
+        let mut capability = Some(capabilities.into_element());
+        let mut buffer = Vec::new();
+        move |frontiers| {
+            let mut output_handle_core = output_buf.activate();
+            let mut output_buf = ConsolidateBuffer::new(&mut output_handle_core, 0);
+            input.for_each(|input_cap, data| {
+                // Note that we can't use `input_cap` to get an output session because we might have advanced our output
+                // frontier already beyond the frontier of this capability.
+
+                // `capability` will be None if we are past the last refresh. We have made sure to not receive any
+                // data that is after the last refresh by setting the `until` of the dataflow to the last refresh.
+                let capability = capability
+                    .as_mut()
+                    .expect("should have a capability if we received data");
+
+                data.swap(&mut buffer);
+                let mut cached_ts: Option<Timestamp> = None;
+                let mut cached_rounded_up_data_ts = None;
+                for (_d, ts, _r) in buffer.iter_mut() {
+                    let rounded_up_data_ts = {
+                        // We cache the rounded up timestamp for the last seen timestamp,
+                        // because the rounding up has a non-negligible cost. Caching for
+                        // just the 1 last timestamp helps already, because in some common
+                        // cases, we'll be seeing a lot of identical timestamps, e.g.,
+                        // during a rehydration, or when we have much more than 1000 records
+                        // during a single second.
+                        if cached_ts != Some(*ts) {
+                            cached_ts = Some(*ts);
+                            cached_rounded_up_data_ts = refresh_schedule.round_up_timestamp(*ts);
+                        }
+                        cached_rounded_up_data_ts
+                    };
+                    match rounded_up_data_ts {
+                        Some(rounded_up_ts) => {
+                            *ts = rounded_up_ts;
+                        }
+                        None => {
+                            // This record is after the last refresh, which is not possible because we set the dataflow
+                            // `until` to the last refresh.
+                            unreachable!("Received data after the last refresh");
```

Comment:
Same here, can we handle this error without crashing?

Reply:
Changed to `soft_panic_or_log`